sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ManifestFileCommitProtocol.scala

```diff
@@ -70,7 +81,29 @@ class ManifestFileCommitProtocol(jobId: String, path: String)
 
   override def abortJob(jobContext: JobContext): Unit = {
     require(fileLog != null, "setupManifestOptions must be called before this function")
-    // Do nothing
+    // Best effort cleanup of complete files from failed job.
+    // Since the file has UUID in its filename, we are safe to try deleting them
+    // as the file will not conflict with file with another attempt on same task.
+    if (pendingCommitFiles.nonEmpty) {
+      pendingCommitFiles.foreach { file =>
+        try {
+          val path = new Path(file)
+          val fs = path.getFileSystem(jobContext.getConfiguration)
+          // this is to make sure the file can be seen from driver as well
+          if (fs.exists(path)) {
+            fs.delete(path, false)
+          }
+        } catch {
+          case e: IOException =>
+            logWarning(s"Fail to remove temporary file $file, continue removing next.", e)
+        }
+      }
+      pendingCommitFiles.clear()
+    }
+  }
+
+  override def onTaskCommit(taskCommit: TaskCommitMessage): Unit = {
+    pendingCommitFiles ++= taskCommit.obj.asInstanceOf[Seq[SinkFileStatus]].map(_.path)
```

Comment:
Just a question. When we handle `SinkFileStatus`, we hit character escaping issue frequently before.
So, in `SinkFileStatus`, we use `new Path(new URI(path))`.

Do we have a test coverage for this newly add code in this PR? I'm worrying about line 90 and line 106.

Reply:
Nice finding. Thanks! I agree we need to use same approach here to ensure path will be read same between reading (file source) and cleaning up (here). Updated existing test to contain special char in output directory as well.